File: src/main/scala/Advent2020/Day15/Day15.scala
package Advent2020.Day15

import Shared.Base

object Day15 extends Base with App {

  val sz = 30000000

  val starting = firstLine.split(",").map(_.toInt)

  def gen(s: Array[Int] = starting, end: Int = 2020) = {
    val ages = Array.fill(sz)(0)
    s.indices foreach { i => ages(s(i)) = i }
    var x = s.length
    ages(0) = x
    var y = 0
    s.length + 1 until end - 1 foreach { i =>
      y = ages(x)
      ages(x) = i
      x = if (y == 0) 0 else i - y
    }
    x
  }

  println(gen())

  time(println(gen(end=sz)))
}

File: src/main/scala/Advent2020/Day15/Day15.js
ls = [0,8,15,2,12,1,4]
sz = 30000000

ages = new Array(sz).fill(0)

function gen(s=ls, end=2020) {
  let i;
  for (i=0;i<s.length;++i) ages[s[i]] = i
  let x = s.length
  ages[0] = x
  let y = ages[x]
  for (i=x+1; i < end-1; ++i) {
    ages[x] = i
    x = y !== 0 ? i - y : 0
    y = ages[x]
  }
  return x
}

function test() {
  t0 = performance.now()
  console.log(gen(ls,sz))
  console.log("Took: " + (performance.now() - t0)/1e3)
  ages = new Array(sz).fill(0)
}

test()